geomstats/stiefel.py

```diff
@@ -0,0 +1,351 @@
+                tangent_mean = tangent_mean + weight_i * self.log(
+                                                    point=point_i,
+                                                    base_point=mean)
+            tangent_mean /= sum_weights
+
+            mean_next = self.exp(
+                tangent_vec=tangent_mean,
+                base_point=mean)
+
+            sq_dist = self.squared_dist(mean_next, mean)
+            sq_dists_between_iterates.append(sq_dist)
+
+            variance = self.variance(points=points,
+                                     weights=weights,
+                                     base_point=mean_next)
+            if gs.isclose(variance, 0):
+                break
+            if sq_dist <= epsilon * variance:
+                break
+
+            mean = mean_next
+            iteration += 1
+
+        if iteration is n_max_iterations:
+            print('Maximum number of iterations {} reached.'
+                  'The mean may be inaccurate'.format(n_max_iterations))
+        return mean
+
+    def exp(self, tangent_vec, base_point):
+        """
+        Riemannian exponential of a tangent vector wrt to a base point.
+        """
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=3)
+        n_tangent_vecs, _, _ = tangent_vec.shape
+
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        n_base_points, n, p = base_point.shape
+
+        assert (n_tangent_vecs == n_base_points
+                or n_tangent_vecs == 1
+                or n_base_points == 1)
+
+        A = gs.matmul(gs.transpose(base_point, axes=(0, 2, 1)), tangent_vec)
+        K = tangent_vec - gs.matmul(base_point, A)
+
+        Q = gs.zeros((K.shape))
+        R = gs.zeros((K.shape[0], K.shape[2], K.shape[2]))
+        for i, k in enumerate(K):
+            Q[i], R[i] = gs.linalg.qr(k)
+
+        AR = gs.concatenate([A, -gs.transpose(R, axes=(0, 2, 1))], axis=2)
+        RZ = gs.concatenate([R, gs.zeros((n_base_points, p, p))], axis=2)
+        block = gs.concatenate([AR, RZ], axis=1)
+        MNe = gs.expm(block)
+
+        return gs.matmul(
+            gs.concatenate([base_point, Q], axis=2), MNe[:, :, 0:p])
+
+    def log(self, point, base_point, max_iter=100, tol=1e-6):
+        """
+        Riemannian logarithm of a point wrt a base point.
+
+        Based on:
+        Zimmermann, Ralf
+        "A Matrix-Algebraic Algorithm for the Riemannian Logarithm
+        on the Stiefel Manifold under the Canonical Metric"
+        SIAM J. Matrix Anal. & Appl., 38(2), 322–342.
+        """
+        point = gs.to_ndarray(point, to_ndim=3)
+        n_points, _, _ = point.shape
+
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        n_base_points, n, p = base_point.shape
+
+        assert (n_points == n_base_points
+                or n_points == 1
+                or n_base_points == 1)
+
+        M = gs.matmul(gs.transpose(base_point, (0, 2, 1)), point)
+
+        # QR of normal component of a point
```

Comment:
Avoid comment within the code as we often forget to modify it while modifying the code and it soon makes no sense anymore. Put it in the docstring? Thank you!